File: packages/config-extends/src/resolve.ts
import path from "node:path";
import {
  getTemplatePartialNames,
  renderTemplate,
  resolveTemplatePartials
} from "toolcraft-design";
import { findBase } from "./discover.js";
import { hasOwnErrorCode } from "./error-codes.js";
import { mergeLayers } from "./merge.js";
import { parseDocument } from "./parse.js";
import type {
  BaseLayer,
  ChainLayer,
  DataLayer,
  DocumentLayer,
  ResolveOptions,
  ResolvedDocument
} from "./types.js";

const MAX_EXTENDS_DEPTH = 5;
const YIELD_TOKEN = "{{yield}}";

interface ClassifiedChain {
  baseLayers: BaseLayer[];
  documentIndex: number;
  documentLayer: DocumentLayer;
}

interface ResolvedBaseChain {
  chain: string[];
  layers: DataLayer[];
}

interface ResolvedBase {
  content: string;
  filePath: string;
  source: string;
  matchedBaseIndex?: number;
}

export async function resolve(
  chain: ChainLayer[],
  options: ResolveOptions
): Promise<ResolvedDocument> {
  const { baseLayers, documentIndex, documentLayer } = classifyChain(chain);
  const parsedDocument = parseDocument(documentLayer.content, documentLayer.filePath);
  const resolvedBase = shouldResolveBase(parsedDocument, options.autoExtend)
    ? await resolveBaseChain({
        name: documentLayer.baseName ?? getBaseName(documentLayer.filePath),
        baseLayers,
        extendsValue: parsedDocument.extends,
        fromFilePath: documentLayer.filePath,
        options,
        optional: !parsedDocument.extends,
        visited: new Set([path.resolve(documentLayer.filePath)]),
        depth: 1
      })
    : undefined;
  const promptFiles = [documentLayer.filePath, ...(resolvedBase?.chain ?? [])];
  const expandedPrompts = await expandPromptPartials(
    {
      source: documentLayer.source,
      data: parsedDocument.data
    },
    resolvedBase?.layers ?? [],
    promptFiles,
    options
  );
  const composedPrompt = composePromptChain(
    expandedPrompts.documentLayer,
    expandedPrompts.baseLayers
  );
  const renderedPrompt =
    composedPrompt === undefined ? undefined : renderPrompt(composedPrompt.prompt, options);
  const merged = mergeLayers([
    ...collectDataLayers(chain.slice(0, documentIndex)),
    {
      source: documentLayer.source,
      data: withResolvedPrompt(parsedDocument.data, renderedPrompt)
    },
    ...stripResolvedBasePrompts(
      resolvedBase?.layers ?? [],
      composedPrompt?.consumedBaseIndexes ?? new Set<number>()
    ),
    ...collectDataLayers(chain.slice(documentIndex + 1))
  ]);

  if (
    composedPrompt !== undefined &&
    getOwnEntry(merged.sources, "prompt") === documentLayer.source &&
    composedPrompt.source !== undefined
  ) {
    merged.sources.prompt = composedPrompt.source;
  }

  return {
    data: merged.data,
    sources: merged.sources,
    chain: [...promptFiles, ...expandedPrompts.partialFiles]
  };
}

function classifyChain(chain: ChainLayer[]): ClassifiedChain {
  const baseLayers: BaseLayer[] = [];
  const documentLayers: Array<{ index: number; layer: DocumentLayer }> = [];

  for (const [index, layer] of chain.entries()) {
    if (isDataLayer(layer)) {
      continue;
    }

    if (isDocumentLayer(layer)) {
      documentLayers.push({ index, layer });
      continue;
    }

    if (isBaseLayer(layer)) {
      baseLayers.push(layer);
    }
  }

  if (documentLayers.length !== 1) {
    throw new Error(`Exactly one document layer is required, received ${documentLayers.length}.`);
  }

  return {
    baseLayers,
    documentIndex: documentLayers[0].index,
    documentLayer: documentLayers[0].layer
  };
}

async function resolveBaseChain({
  name,
  baseLayers,
  extendsValue,
  fromFilePath,
  options,
  optional,
  visited,
  depth
}: {
  name: string;
  baseLayers: BaseLayer[];
  extendsValue: boolean | string;
  fromFilePath: string;
  options: ResolveOptions;
  optional: boolean;
  visited: Set<string>;
  depth: number;
}): Promise<ResolvedBaseChain | undefined> {
  if (depth > MAX_EXTENDS_DEPTH) {
    throw new Error(`Maximum extends depth exceeded (${MAX_EXTENDS_DEPTH}).`);
  }

  const discoveredBase = await resolveBase({
    name,
    baseLayers,
    extendsValue,
    fromFilePath,
    options,
    optional
  });

  if (discoveredBase === undefined) {
    return undefined;
  }

  const resolvedBasePath = path.resolve(discoveredBase.filePath);

  if (visited.has(resolvedBasePath)) {
    if (optional) {
      return undefined;
    }

    throw new Error(
      `Circular extends detected.\nVisited files:\n- ${[...visited, resolvedBasePath].join("\n- ")}`
    );
  }

  const parsedBase = parseDocument(discoveredBase.content, discoveredBase.filePath);
  const nextVisited = new Set(visited);
  nextVisited.add(resolvedBasePath);
  const nestedBase = parsedBase.extends
    ? await resolveBaseChain({
        name: getBaseName(discoveredBase.filePath),
        baseLayers:
          discoveredBase.matchedBaseIndex === undefined
            ? baseLayers
            : baseLayers.slice(discoveredBase.matchedBaseIndex + 1),
        extendsValue: parsedBase.extends,
        fromFilePath: discoveredBase.filePath,
        options,
        optional: false,
        visited: nextVisited,
        depth: depth + 1
      })
    : undefined;

  return {
    layers: [
      {
        source: discoveredBase.source,
        data: parsedBase.data
      },
      ...(nestedBase?.layers ?? [])
    ],
    chain: [discoveredBase.filePath, ...(nestedBase?.chain ?? [])]
  };
}

async function resolveBase({
  name,
  baseLayers,
  extendsValue,
  fromFilePath,
  options,
  optional
}: {
  name: string;
  baseLayers: BaseLayer[];
  extendsValue: boolean | string;
  fromFilePath: string;
  options: ResolveOptions;
  optional: boolean;
}): Promise<ResolvedBase | undefined> {
  if (typeof extendsValue === "string") {
    return readPathValuedBase(extendsValue, fromFilePath, options);
  }

  let discoveredBase;

  try {
    discoveredBase = await findBase(
      name,
      baseLayers.map((layer) => layer.path),
      options.fs
    );
  } catch (error) {
    if (optional && isBaseNotFoundError(error)) {
      return undefined;
    }

    throw error;
  }

  const discoveredBaseDirectory = path.resolve(path.dirname(discoveredBase.filePath));
  const matchedBaseIndex = baseLayers.findIndex(
    (layer) => path.resolve(layer.path) === discoveredBaseDirectory
  );

  if (matchedBaseIndex === -1) {
    throw new Error(`Resolved base is outside configured base paths: ${discoveredBase.filePath}`);
  }

  return {
    content: discoveredBase.content,
    filePath: discoveredBase.filePath,
    matchedBaseIndex,
    source: baseLayers[matchedBaseIndex].source
  };
}

async function readPathValuedBase(
  extendsPath: string,
  fromFilePath: string,
  options: ResolveOptions
): Promise<ResolvedBase> {
  const filePath = path.resolve(path.dirname(fromFilePath), extendsPath);

  try {
    return {
      content: await options.fs.readFile(filePath, "utf8"),
      filePath,
      source: filePath
    };
  } catch (error) {
    if (hasOwnErrorCode(error, "ENOENT") || hasOwnErrorCode(error, "ENOTDIR")) {
      throw new Error(`base file not found at ${filePath}`);
    }

    throw error;
  }
}

function collectDataLayers(chain: ChainLayer[]): DataLayer[] {
  return chain.filter(isDataLayer);
}

interface ComposedPromptResult {
  consumedBaseIndexes: Set<number>;
  prompt: string;
  source?: string;
}

function composePromptChain(
  documentLayer: DataLayer,
  baseLayers: DataLayer[]
): ComposedPromptResult | undefined {
  const documentPrompt = getOwnEntry(documentLayer.data, "prompt");

  if (documentPrompt !== undefined && typeof documentPrompt !== "string") {
    return undefined;
  }

  if (documentPrompt !== undefined) {
    assertValidYieldCount(documentPrompt);
  }

  let prompt = documentPrompt;
  let source = prompt === undefined || prompt === "" ? undefined : documentLayer.source;
  const consumedBaseIndexes = new Set<number>();

  for (const [index, layer] of baseLayers.entries()) {
    const candidate = getOwnEntry(layer.data, "prompt");

    if (candidate === undefined) {
      continue;
    }

    if (typeof candidate !== "string") {
      break;
    }

    assertValidYieldCount(candidate);
    consumedBaseIndexes.add(index);
    prompt = composeAdjacentPrompts(prompt, candidate);

    if (source === undefined && candidate !== "") {
      source = layer.source;
    }
  }

  if (prompt !== undefined && prompt.includes(YIELD_TOKEN)) {
    throw new Error('Final resolved prompt contains an unresolved "{{yield}}" token.');
  }

  if (prompt === undefined) {
    return undefined;
  }

  return {
    consumedBaseIndexes,
    prompt,
    source
  };
}

function composeAdjacentPrompts(high: string | undefined, low: string): string {
  if (high === undefined || high === "") {
    return low.includes(YIELD_TOKEN) ? replaceYield(low, "") : low;
  }

  if (high.includes(YIELD_TOKEN)) {
    return replaceYield(high, low);
  }

  if (low.includes(YIELD_TOKEN)) {
    return replaceYield(low, high);
  }

  return high;
}

function replaceYield(prompt: string, replacement: string): string {
  return prompt.split(YIELD_TOKEN).join(replacement);
}

async function expandPromptPartials(
  documentLayer: DataLayer,
  baseLayers: DataLayer[],
  promptFiles: string[],
  options: ResolveOptions
): Promise<{ baseLayers: DataLayer[]; documentLayer: DataLayer; partialFiles: string[] }> {
  const directories = unique(promptFiles.map((filePath) => path.dirname(filePath)));
  const partials: Record<string, string> = Object.create(null) as Record<string, string>;
  const partialFiles: string[] = [];

  const loadPartial = async (name: string): Promise<void> => {
    if (name.trim().length === 0) {
      throw new Error("Partial name must be non-empty.");
    }

    if (Object.hasOwn(partials, name)) {
      return;
    }

    const partial = await findPartial(name, directories, options.fs);
    partials[name] = partial.content;
    partialFiles.push(partial.filePath);
    for (const nestedName of getTemplatePartialNames(partial.content)) {
      await loadPartial(nestedName);
    }
  };

  const promptLayers = [documentLayer, ...baseLayers];
  for (const layer of promptLayers) {
    const prompt = getOwnEntry(layer.data, "prompt");
    if (typeof prompt !== "string") {
      continue;
    }
    for (const name of getTemplatePartialNames(prompt)) {
      await loadPartial(name);
    }
  }

  const expandedLayers = promptLayers.map((layer) => withExpandedPrompt(layer, partials));

  return {
    documentLayer: expandedLayers[0],
    baseLayers: expandedLayers.slice(1),
    partialFiles
  };
}

function withExpandedPrompt(layer: DataLayer, partials: Record<string, string>): DataLayer {
  const prompt = getOwnEntry(layer.data, "prompt");
  if (typeof prompt !== "string") {
    return layer;
  }

  return {
    source: layer.source,
    data: {
      ...layer.data,
      prompt: resolveTemplatePartials(prompt, partials)
    }
  };
}

function renderPrompt(prompt: string, options: ResolveOptions): string {
  if (options.view === undefined && options.validate !== true) {
    return prompt;
  }

  return renderTemplate(prompt, options.view ?? {}, { escape: "none", validate: options.validate });
}

async function findPartial(
  name: string,
  directories: string[],
  fs: ResolveOptions["fs"]
): Promise<{ content: string; filePath: string }> {
  const checkedPaths: string[] = [];

  for (const directory of directories) {
    const filePath = path.join(directory, `${name}.md`);
    assertInsideDirectory(name, directory, filePath);
    checkedPaths.push(filePath);

    try {
      return { content: await fs.readFile(filePath, "utf8"), filePath };
    } catch (error) {
      if (hasOwnErrorCode(error, "ENOENT")) {
        continue;
      }
      throw error;
    }
  }

  throw new Error(`Partial "${name}" not found.\nChecked paths:\n- ${checkedPaths.join("\n- ")}`);
}

function assertInsideDirectory(name: string, directory: string, filePath: string): void {
  const relativePath = path.relative(path.resolve(directory), path.resolve(filePath));
  if (
    relativePath === ".." ||
    relativePath.startsWith(`..${path.sep}`) ||
    path.isAbsolute(relativePath)
  ) {
    throw new Error(`Partial name must remain inside prompt directories: "${name}".`);
  }
}

function unique(values: string[]): string[] {
  return [...new Set(values)];
}

function assertValidYieldCount(prompt: string): void {
  if (countYieldTokens(prompt) > 1) {
    throw new Error('Prompt composition supports exactly one "{{yield}}" token per prompt.');
  }
}

function countYieldTokens(prompt: string): number {
  return prompt.split(YIELD_TOKEN).length - 1;
}

function withResolvedPrompt(
  data: Record<string, unknown>,
  prompt: string | undefined
): Record<string, unknown> {
  if (prompt === undefined) {
    return data;
  }

  return {
    ...data,
    prompt
  };
}

function stripResolvedBasePrompts(
  layers: DataLayer[],
  consumedBaseIndexes: Set<number>
): DataLayer[] {
  return layers.map((layer, index) => {
    if (!consumedBaseIndexes.has(index) || typeof getOwnEntry(layer.data, "prompt") !== "string") {
      return layer;
    }

    const { prompt: ignoredPrompt, ...data } = layer.data;

    void ignoredPrompt;

    return {
      source: layer.source,
      data
    };
  });
}

function getBaseName(filePath: string): string {
  return path.basename(filePath, path.extname(filePath));
}

function shouldResolveBase(
  parsedDocument: ReturnType<typeof parseDocument>,
  autoExtend: boolean | undefined
): boolean {
  return parsedDocument.extends !== false || (autoExtend === true && !parsedDocument.hasExtendsField);
}

function getOwnEntry(record: Record<string, unknown>, key: string): unknown {
  return Object.prototype.hasOwnProperty.call(record, key) ? record[key] : undefined;
}

function isBaseNotFoundError(error: unknown): error is Error {
  return (
    error instanceof Error &&
    error.message.startsWith('Base "') &&
    error.message.includes('" not found.\nChecked paths:')
  );
}

function isDataLayer(layer: ChainLayer): layer is DataLayer {
  return "data" in layer;
}

function isDocumentLayer(layer: ChainLayer): layer is DocumentLayer {
  return "filePath" in layer && "content" in layer;
}

function isBaseLayer(layer: ChainLayer): layer is BaseLayer {
  return "path" in layer;
}

File: packages/config-extends-rust/src/resolve.rs
//! Owned document resolution with host filesystem/path operations and own templates.
use crate::{
    Layer,
    discover::{self, Error},
    document::{self, Extends},
    merge_layers, prompt,
};
use config_mutations_rust::value::Value;
use toolcraft_design_rust::{
    data::{DataEnvironment, DataHost, Graph, Node},
    template::{self, Partials, RenderOptions},
};
#[derive(Clone, Debug)]
pub struct DocumentLayer {
    pub source: Vec<u16>,
    pub file_path: Vec<u16>,
    pub content: Vec<u16>,
    pub base_name: Option<Vec<u16>>,
}
#[derive(Clone, Debug)]
pub struct BaseLayer {
    pub source: Vec<u16>,
    pub path: Vec<u16>,
}
#[derive(Clone, Debug)]
pub enum ChainLayer {
    Data(Layer),
    Document(DocumentLayer),
    Base(BaseLayer),
}
#[derive(Default)]
pub struct Options<'a> {
    pub auto_extend: bool,
    pub validate: bool,
    pub view: Option<&'a Graph>,
}
#[derive(Debug)]
pub struct Resolved {
    pub data: Value,
    pub sources: Vec<(Vec<u16>, Vec<u16>)>,
    pub chain: Vec<Vec<u16>>,
}
/// Document/base layers prepared before foreign data layers are inspected.
#[derive(Debug)]
pub struct Prepared {
    pub document_index: usize,
    pub layers: Vec<Layer>,
    pub document_source: Vec<u16>,
    pub prompt_source: Option<Vec<u16>>,
    pub chain: Vec<Vec<u16>>,
}
pub trait Host: discover::Host {
    fn resolve(&mut self, path: &[u16]) -> Vec<u16>;
    fn dirname(&mut self, path: &[u16]) -> Vec<u16>;
    /// Basename with its extension removed.
    fn basename(&mut self, path: &[u16]) -> Vec<u16>;
    /// Lowercase extension, including its leading dot.
    fn extension(&mut self, path: &[u16]) -> Vec<u16>;
    fn is_absolute(&mut self, path: &[u16]) -> bool;
    /// Admit a document using the own parser; foreign hosts may preserve runtime values.
    fn admit(
        &mut self,
        content: &[u16],
        file: &[u16],
    ) -> Result<document::ParsedDocument, Error<Self::Error>> {
        let extension = self.extension(file);
        document::parse_document(
            content,
            &extension,
            file,
            &mut |path| self.is_absolute(path),
            None,
        )
        .map_err(|error| Error::Policy(error.message))
    }
    /// Render with the own data engine; foreign hosts may provide runtime views.
    fn render(
        &mut self,
        prompt: &[u16],
        options: &Options<'_>,
    ) -> Result<Vec<u16>, Error<Self::Error>> {
        let empty = Graph {
            nodes: vec![Node::Object(vec![])],
            root: 0,
        };
        let graph = options.view.unwrap_or(&empty);
        let mut partials = NoPartials;
        let mut environment = DataEnvironment::new(graph, &mut partials).map_err(template_error)?;
        template::render(
            prompt,
            0,
            &mut environment,
            RenderOptions {
                escape_none: true,
                validate: options.validate,
                ..RenderOptions::default()
            },
        )
        .map_err(template_error)
    }
    /// Own ENOTDIR is treated as missing only for path-valued extends.
    fn path_not_found(&self, _error: &Self::Error) -> bool {
        false
    }
}
fn u(text: &str) -> Vec<u16> {
    text.encode_utf16().collect()
}
fn policy<E>(text: &str) -> Error<E> {
    Error::Policy(u(text))
}
fn named<E>(prefix: &str, name: &[u16], suffix: &str) -> Error<E> {
    let mut message = u(prefix);
    message.extend(name);
    message.extend(u(suffix));
    Error::Policy(message)
}
fn set_prompt(layer: &mut Layer, prompt: Vec<u16>) {
    let Value::Object(fields) = &mut layer.data else {
        unreachable!("Document layers have object roots")
    };
    if let Some((_, value)) = fields.iter_mut().find(|(key, _)| *key == u("prompt")) {
        *value = Value::String(prompt);
    } else {
        fields.push((u("prompt"), Value::String(prompt)));
    }
}
type BaseChain = (Vec<Layer>, Vec<Vec<u16>>);
type Partial = (Vec<u16>, Vec<u16>);
async fn load_bases<H: Host>(
    document: &DocumentLayer,
    extends: Extends,
    bases: &[BaseLayer],
    optional: bool,
    host: &mut H,
) -> Result<BaseChain, Error<H::Error>> {
    let mut layers = vec![];
    let mut files = vec![];
    let mut visited = vec![host.resolve(&document.file_path)];
    let mut name = document
        .base_name
        .clone()
        .unwrap_or_else(|| host.basename(&document.file_path));
    let mut from = document.file_path.clone();
    let mut value = extends;
    let mut optional = optional;
    let mut start = 0;
    for depth in 1.. {
        if depth > 5 {
            return Err(policy("Maximum extends depth exceeded (5)."));
        }
        let (content, file, source, next_start) = match value {
            Extends::Path(ref path) => {
                let directory = host.dirname(&from);
                let joined = host.join(&directory, path);
                let file = host.resolve(&joined);
                let content = match host.read(&file).await {
                    Ok(Some(content)) => content,
                    Ok(None) => return Err(named("base file not found at ", &file, "")),
                    Err(error) if host.path_not_found(&error) => {
                        return Err(named("base file not found at ", &file, ""));
                    }
                    Err(error) => return Err(Error::Host(error)),
                };
                (content, file.clone(), file, start)
            }
            _ => {
                let paths: Vec<_> = bases[start..]
                    .iter()
                    .map(|base| base.path.clone())
                    .collect();
                let discovered = match discover::find_base(&name, &paths, host).await {
                    Ok(discovered) => discovered,
                    Err(Error::Policy(message))
                        if optional && message.starts_with(&u("Base \"")) =>
                    {
                        break;
                    }
                    Err(error) => return Err(error),
                };
                let directory = host.dirname(&discovered.file_path);
                let directory = host.resolve(&directory);
                let index = bases[start..]
                    .iter()
                    .position(|base| host.resolve(&base.path) == directory)
                    .map(|index| index + start)
                    .ok_or_else(|| {
                        named(
                            "Resolved base is outside configured base paths: ",
                            &discovered.file_path,
                            "",
                        )
                    })?;
                (
                    discovered.content,
                    discovered.file_path,
                    bases[index].source.clone(),
                    index + 1,
                )
            }
        };
        let resolved = host.resolve(&file);
        if visited.contains(&resolved) {
            if optional {
                break;
            }
            let mut message = u("Circular extends detected.\nVisited files:\n- ");
            for (index, path) in visited.iter().chain(std::iter::once(&resolved)).enumerate() {
                if index > 0 {
                    message.extend(u("\n- "));
                }
                message.extend(path);
            }
            return Err(Error::Policy(message));
        }
        let parsed = host.admit(&content, &file)?;
        visited.push(resolved);
        files.push(file.clone());
        layers.push(Layer {
            source,
            data: parsed.yaml.value,
        });
        if parsed.extends == Extends::Disabled {
            break;
        }
        value = parsed.extends;
        name = host.basename(&file);
        from = file;
        optional = false;
        start = next_start;
    }
    Ok((layers, files))
}
async fn partial<H: Host>(
    name: &[u16],
    directories: &[Vec<u16>],
    host: &mut H,
) -> Result<Partial, Error<H::Error>> {
    let mut checked = vec![];
    for directory in directories {
        let mut file = name.to_vec();
        file.extend(u(".md"));
        let path = host.join(directory, &file);
        if !host.contains(directory, &path) {
            return Err(named(
                "Partial name must remain inside prompt directories: \"",
                name,
                "\".",
            ));
        }
        checked.push(path.clone());
        if let Some(content) = host.read(&path).await.map_err(Error::Host)? {
            return Ok((content, path));
        }
    }
    let mut message = u("Partial \"");
    message.extend(name);
    message.extend(u("\" not found.\nChecked paths:\n- "));
    for (index, path) in checked.iter().enumerate() {
        if index > 0 {
            message.extend(u("\n- "));
        }
        message.extend(path);
    }
    Err(Error::Policy(message))
}
fn template_error<E>(error: template::Error) -> Error<E> {
    Error::Policy(error.description)
}
async fn expand<H: Host>(
    document: &mut Layer,
    bases: &mut [Layer],
    files: &[Vec<u16>],
    host: &mut H,
) -> Result<Vec<Vec<u16>>, Error<H::Error>> {
    let mut directories = vec![];
    for file in files {
        let dir = host.dirname(file);
        if !directories.contains(&dir) {
            directories.push(dir);
        }
    }
    let mut partials: Vec<(Vec<u16>, Vec<u16>)> = vec![];
    let mut partial_files = vec![];
    for layer in std::iter::once(&*document).chain(bases.iter()) {
        let Some(Value::String(prompt)) = layer.data.get("prompt") else {
            continue;
        };
        let mut pending = template::partial_names(prompt)
            .map_err(template_error)?
            .into_iter()
            .rev()
            .collect::<Vec<_>>();
        while let Some(name) = pending.pop() {
            if name.iter().all(|unit|matches!(unit,9..=13|32|160|0xfeff|0x1680|0x2000..=0x200a|0x2028|0x2029|0x202f|0x205f|0x3000)){return Err(policy("Partial name must be non-empty."));}
            if partials.iter().any(|(key, _)| *key == name) {
                continue;
            }
            let (content, file) = partial(&name, &directories, host).await?;
            let nested = template::partial_names(&content).map_err(template_error)?;
            partials.push((name, content));
            partial_files.push(file);
            pending.extend(nested.into_iter().rev());
        }
    }
    for layer in std::iter::once(document).chain(bases.iter_mut()) {
        if let Some(Value::String(prompt)) = layer.data.get("prompt") {
            let prompt = template::expand_partials(prompt, &partials).map_err(template_error)?;
            set_prompt(layer, prompt);
        }
    }
    Ok(partial_files)
}
struct NoPartials;
impl Partials for NoPartials {
    fn has(&mut self, _name: &[u16]) -> Result<bool, template::Error> {
        Ok(false)
    }
    fn get(&mut self, name: &[u16]) -> Result<Vec<u16>, template::Error> {
        let mut description = u("Partial \"");
        description.extend(name);
        description.extend(u("\" not found."));
        Err(template::Error {
            description,
            line: None,
            column: None,
        })
    }
}
impl DataHost for NoPartials {}
pub async fn prepare<H: Host>(
    chain: &[ChainLayer],
    options: &Options<'_>,
    host: &mut H,
) -> Result<Prepared, Error<H::Error>> {
    let documents: Vec<_> = chain
        .iter()
        .enumerate()
        .filter_map(|(index, layer)| match layer {
            ChainLayer::Document(document) => Some((index, document)),
            _ => None,
        })
        .collect();
    if documents.len() != 1 {
        return Err(policy(&format!(
            "Exactly one document layer is required, received {}.",
            documents.len()
        )));
    }
    let (index, document) = documents[0];
    let bases: Vec<_> = chain
        .iter()
        .filter_map(|layer| match layer {
            ChainLayer::Base(base) => Some(base.clone()),
            _ => None,
        })
        .collect();
    let parsed = host.admit(&document.content, &document.file_path)?;
    let should_extend =
        parsed.extends != Extends::Disabled || options.auto_extend && !parsed.has_extends;
    let (mut base_layers, base_files) = if should_extend {
        let optional = parsed.extends == Extends::Disabled;
        load_bases(document, parsed.extends, &bases, optional, host).await?
    } else {
        (vec![], vec![])
    };
    let mut prompt_files = vec![document.file_path.clone()];
    prompt_files.extend(base_files);
    let mut expanded_document = Layer {
        source: document.source.clone(),
        data: parsed.yaml.value,
    };
    let partial_files = expand(
        &mut expanded_document,
        &mut base_layers,
        &prompt_files,
        host,
    )
    .await?;
    let composed = prompt::compose_prompts(&expanded_document, &base_layers).map_err(policy)?;
    if let Some(composed) = &composed {
        let prompt = if options.view.is_some() || options.validate {
            host.render(&composed.prompt, options)?
        } else {
            composed.prompt.clone()
        };
        set_prompt(&mut expanded_document, prompt);
        for index in &composed.consumed {
            let Value::Object(fields) = &mut base_layers[*index].data else {
                unreachable!()
            };
            fields.retain(|(key, _)| *key != u("prompt"));
        }
    }
    let mut layers = vec![expanded_document];
    layers.extend(base_layers);
    prompt_files.extend(partial_files);
    Ok(Prepared {
        document_index: index,
        layers,
        document_source: document.source.clone(),
        prompt_source: composed.and_then(|prompt| prompt.source),
        chain: prompt_files,
    })
}
/// Resolve owned layers after document preparation; foreign runtimes can defer their merge.
pub async fn resolve<H: Host>(
    chain: &[ChainLayer],
    options: &Options<'_>,
    host: &mut H,
) -> Result<Resolved, Error<H::Error>> {
    let prepared = prepare(chain, options, host).await?;
    let data_layers = |layers: &[ChainLayer]| {
        layers
            .iter()
            .filter_map(|layer| match layer {
                ChainLayer::Data(data) => Some(data.clone()),
                _ => None,
            })
            .collect::<Vec<_>>()
    };
    let mut layers = data_layers(&chain[..prepared.document_index]);
    layers.extend(prepared.layers);
    layers.extend(data_layers(&chain[prepared.document_index + 1..]));
    let mut merged = merge_layers(&layers).map_err(policy)?;
    if let Some(source) = prepared.prompt_source
        && let Some((_, provenance)) = merged
            .sources
            .iter_mut()
            .find(|(key, value)| *key == u("prompt") && *value == prepared.document_source)
    {
        *provenance = source;
    }
    Ok(Resolved {
        data: merged.data,
        sources: merged.sources,
        chain: prepared.chain,
    })
}
